File: 36_Kadane_Algorithm/Java/KadaneAlgo.java
// Optimised.java

public class KadaneAlgo {
    public static int maxSubarraySum(int[] arr) {
        int n = arr.length;

        int maxEndingHere = arr[0];
        int maxSoFar = arr[0];

        for (int i = 1; i < n; i++) {
            // Either extend the previous subarray or start a new one
            maxEndingHere = Math.max(arr[i], maxEndingHere + arr[i]);
            maxSoFar = Math.max(maxSoFar, maxEndingHere);
        }

        return maxSoFar;
    }

    public static void main(String[] args) {
        int[] arr = {2, 3, -8, 7, -1, 2, 3};
        System.out.println("Maximum Subarray Sum (Kadane's Algorithm): " + maxSubarraySum(arr));
    }
}

/*
Time Complexity: O(n)
Space Complexity: O(1)
*/

File: Arrays/Kadane_Algorithm/Python/KadaneAlgo.py


def max_subarray_sum_kadane(arr):
    max_ending_here = arr[0]
    max_so_far = arr[0]

    for i in range(1, len(arr)):
        max_ending_here = max(arr[i], max_ending_here + arr[i])
        max_so_far = max(max_so_far, max_ending_here)

    return max_so_far

# Example usage
arr = [2, 3, -8, 7, -1, 2, 3]
print("Maximum Subarray Sum (Kadane's Algorithm):", max_subarray_sum_kadane(arr))

# Time Complexity: O(n)
# Space Complexity: O(1)
